net/respondd/src/respondd.c

```diff
@@ -41,150 +48,381 @@
-  puts("  -g <ip6>         multicast group, e.g. ff02::2:1001");
-  puts("  -p <int>         port number to listen on");
-  puts("  -i <string>      interface on which the group is joined");
-  puts("  -c <string>      Lua command");
-  puts("  -h               this help\n");
+	puts("Usage: respondd [-h] -g <group> -p <port> -i <if0> [-i <if1> ..]");
+	puts("	-g <ip6>         multicast group, e.g. ff02::2:1001");
+	puts("	-p <int>         port number to listen on");
+	puts("	-i <string>      interface on which the group is joined");
+	puts("	-d <string>      data provider directory");
+	puts("	-h               this help\n");
 }
 
 static void join_mcast(const int sock, const struct in6_addr addr, const char *iface) {
-  struct ipv6_mreq mreq;
+	struct ipv6_mreq mreq;
 
-  mreq.ipv6mr_multiaddr = addr;
-  mreq.ipv6mr_interface = if_nametoindex(iface);
+	mreq.ipv6mr_multiaddr = addr;
+	mreq.ipv6mr_interface = if_nametoindex(iface);
 
-  if (mreq.ipv6mr_interface == 0)
-    goto error;
+	if (mreq.ipv6mr_interface == 0)
+		goto error;
 
-  if (setsockopt(sock, IPPROTO_IPV6, IPV6_JOIN_GROUP, &mreq, sizeof(mreq)) == -1)
-    goto error;
+	if (setsockopt(sock, IPPROTO_IPV6, IPV6_JOIN_GROUP, &mreq, sizeof(mreq)) == -1)
+		goto error;
 
-  return;
+	return;
 
  error:
-  fprintf(stderr, "Could not join multicast group on %s: ", iface);
-  perror(NULL);
+	fprintf(stderr, "Could not join multicast group on %s: ", iface);
+	perror(NULL);
+}
+
+
+static void update_time(void) {
+	struct timespec tp;
+	clock_gettime(CLOCK_MONOTONIC, &tp);
+
+	now = (int64_t)tp.tv_sec * 1000 + tp.tv_nsec / 1000000;
+}
+
+
+/** Merges two JSON objects (on conflicts, object a will be preferred) */
+static struct json_object * merge_json(struct json_object *a, struct json_object *b) {
+	if (!json_object_is_type(a, json_type_object) || !json_object_is_type(b, json_type_object)) {
+		json_object_put(b);
+		return a;
+	}
+
+	json_object_object_foreach(a, key, val_a) {
+		struct json_object *val_b;
+
+		json_object_get(val_a);
+
+		if (!json_object_object_get_ex(b, key, &val_b)) {
+			json_object_object_add(b, key, val_a);
+			continue;
+		}
+
+		json_object_get(val_b);
+
+		json_object_object_add(b, key, merge_json(val_a, val_b));
+	}
+
+	json_object_put(a);
+	return b;
+}
+
+static const struct respondd_provider_info * get_providers(const char *filename) {
+	char path[2 + strlen(filename) + 1];
+	snprintf(path, sizeof(path), "./%s", filename);
+
+	void *handle = dlopen(path, RTLD_NOW|RTLD_LOCAL);
+	if (!handle)
+		return NULL;
+
+	const struct respondd_provider_info *ret = dlsym(handle, "respondd_providers");
+	if (!ret) {
+		dlclose(handle);
+		return NULL;
+	}
+
+	return ret;
+}
+
+static void load_cache_time(struct request_type *r, const char *name) {
+	r->cache_time = 0;
+
+	char filename[strlen(name) + 7];
+	snprintf(filename, sizeof(filename), "%s.cache", name);
+
+	FILE *f = fopen(filename, "r");
+	if (!f)
+		return;
+
+	float val = 0;
+	fscanf(f, "%f", &val);
```

Comment:
I'd like to change statistics and neighbours cache times to something like 0.9 seconds, so a consumer updating the data once a second will always get new data. The new respondd is to much faster than the Lua version that sub-second cache times make actually sense.

We could change this to milliseconds, so no floats are needed. The Lua version used seconds as well, but that hasn't been in any Gluon release yet, so I don't see an issue with changing it.